### PR TITLE
65 user action validation

### LINKED ISSUE #65: Add user access validation to GC action processing
## Description

There needs to be a mechanism that links the user that is performing the action to the action entity that is saved to the DB, and as part of the validation process, checks that the user has the necessary permission to perform that action.

## Dependencies

- with third parties, data etc.

## Tests

- [] What tests are needed to accomplish the success criteria

## Success Criteria

- [] Checkboxes describing every step needed to be taken for defining this ticket as Done.

### LINKED ISSUE #84: Transfer account whitelisting
## Description

- Enable accounts to whitelist other accounts to receive transfers of certificates from

## Tests

- [ ] Transfers from all except whitelisted accounts to target account should fail

## Success Criteria

- [ ] Add attribute to account to list whitelisted IDs of other accounts that can transfer to it
- [ ] For transfer service add check that target transfer account has the transferring account whitelisted

### PR DESCRIPTION
**Pull request by CAGalbraith**
## Description

- Closes #84 : Adds account whitelisting for indicating the account can receive transfers from select accounts 
- Closes #65 : Starts the user role validation function, just looking at user ownership of accounts for now

## Has This Been Tested?

- [ ] Yes

_Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce.
Please also list any relevant details for your test configuration_

## Checklist:

- [ ] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have checked my code and corrected any misspellings

### CODE REVIEW

File: gc_registry/certificate/schemas.py [hunk 1]
```diff
@@ -94,6 +94,11 @@ class GranularCertificateBundleBase(BaseModel):
                         standardised energy volume per Granular Certificate, rounded down to the nearest Wh. Equal to
                         (bundle_id_range_end - bundle_id_range_start + 1)."""
     )
+    beneficiary: str | None = Field(
```
Comment (ClemAtt) [inline gc_registry/certificate/schemas.py:97]: Are these all the changes from the other PR?

File: gc_registry/tests/certificate/test_certificate_services.py [hunk 2]
```diff
@@ -239,6 +239,15 @@ def test_transfer_gcs(
         Transfer a fixed number of certificates from one account to another.
         """
 
+        # Whitelist the source account for the target account
+        fake_db_account_2 = db_write_session.merge(fake_db_account_2)
```
Comment (ClemAtt) [inline gc_registry/tests/certificate/test_certificate_services.py:243]: Nice - do we have test case where not whitelisted to check it fails?

File: gc_registry/account/validation.py [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+    if account_exists:
+        raise HTTPException(
+            status_code=400, detail="Account name already exists in the database."
+        )
+
+    # All user_ids linked to the account must exist in the database
+    user_ids_in_db = (
+        read_session.query(User.id).filter(User.id.in_(account.user_ids)).all()
+    )
+    if len(user_ids_in_db) != len(account.user_ids):
```
Comment (ClemAtt) [inline gc_registry/account/validation.py:28]: Would this check be better a `set` operation to check membership is the same? 

e.g. `set(user_ids_in_db) != set(account.user_ids)`

**Review by ClemAtt — DISMISSED** [review]
Looks good - maybe some minor changes and test changes

File: gc_registry/certificate/schemas.py (hunk 1 above)
Comment (CAGalbraith) [reply gc_registry/certificate/schemas.py:97]: Looks like it - didn't see them in time

File: gc_registry/tests/certificate/test_certificate_services.py (hunk 2 above)
Comment (CAGalbraith) [reply gc_registry/tests/certificate/test_certificate_services.py:243]: Can add that case

File: gc_registry/account/validation.py (hunk 3 above)
Comment (CAGalbraith) [reply gc_registry/account/validation.py:28]: Aye good catch

**Pull request merged**